stan/math/opencl/rev/vari.hpp

```diff
@@ -0,0 +1,190 @@
+#define STAN_MATH_OPENCL_REV_VARI_HPP
+#ifdef STAN_OPENCL
+
+#include <stan/math/rev/core/vari.hpp>
+#include <stan/math/opencl/is_matrix_cl.hpp>
+#include <stan/math/opencl/kernel_generator.hpp>
+
+namespace stan {
+namespace math {
+/**
+ * The variable implementation for `matrix_cl`.
+ *
+ * This class is complete (not abstract) and may be used for
+ * constants.
+ *
+ * A variable implementation is constructed with a constant
+ * value. It also stores the adjoint for storing the partial
+ * derivative with respect to the root of the derivative tree.
+ *
+ */
+template <typename T>
+class vari_value<T, require_kernel_expression_lhs_t<T>>
+    : public vari_base, public chainable_alloc {
+ public:
+  /**
+   * The adjoint of this variable, which is the partial derivative
+   * of this variable with respect to the root variable.
+   */
+  T adj_;
+
+  /**
+   * The value of this variable.
+   */
+  T val_;
+
+  /**
+   * Construct a dense matrix_cl variable implementation from a value. The
+   * adjoint is initialized to zero.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam S A dense Eigen type that is convertible to `value_type`
+   * @param x Value of the constructed variable.
+   */
+  template <typename S, require_convertible_t<S&, T>* = nullptr>
+  explicit vari_value(S&& x)
+      : chainable_alloc(),
+        adj_(constant(0, x.rows(), x.cols())),
+        val_(std::forward<S>(x)) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+ protected:
+  // to allow access to this constructor from instantinations with different
+  // template parameters
+  template <typename, typename>
+  friend class vari_value;
+  /**
+   * Construct a dense matrix_cl variable implementation from a value and
+   * adjoint.
+   *
+   * All constructed variables are added to the stack. Variables
+   * should be constructed before variables on which they depend
+   * to insure proper partial derivative propagation.  During
+   * derivative propagation, the chain() method of each variable
+   * will be called in the reverse order of construction.
+   *
+   * @tparam R A dense Eigen type that is convertible to `T_val`
+   * @tparam S A dense Eigen type that is convertible to `T_adj`
+   * @param val Value of the constructed variable.
+   * @param adj Adjoint of the constructed variable.
+   */
+  template <typename R, typename S, require_convertible_t<R&, T>* = nullptr,
+            require_convertible_t<S&, T>* = nullptr>
+  vari_value(R&& val, S&& adj)
+      : chainable_alloc(),
+        adj_(std::forward<S>(adj)),
+        val_(std::forward<R>(val)) {
+    ChainableStack::instance_->var_stack_.push_back(this);
+  }
+
+ public:
```

Comment:
Is tehre a reason to flip between public and protected here?

Reply:
I used to have everything public, but than you wanted to have this constructor protected or privat. I don't understand what is the issue here.

Reply:
It's not an issue codewise, it's that style wise flipping between access specifiers (public to protected and then back to public) looks a little funny

Reply:
I don't know in what way do you want me to structure this. Back in kernel generator I groupped things so that everything public, private, etc. was together and you wanted me to group things by what is used together. Following that rule I guessed it made sense to put constructors together and that is not ok either. So what rule should I follow?

Reply:
Apologies if I'm being confusing. It's a non-issue but all I mean is right now you have

```cpp
class foo {
public:
 // public stuff
protected:
 // protected stuff
public:
 // more public stuff
}
```

Is there a reason this can't be just one public like this?

```cpp
class foo {
public:
 // public stuff
protected:
 // protected stuff
}
```

Has this shown up previously and I said to do something like multiple access specifiers? My b if so. I'm pretty fine with how it is now but I think it's easier to read if there is just one of each access specifier.

Reply:
Fine, I will change it.

> Has this shown up previously and I said to do something like multiple access specifiers?
https://github.com/stan-dev/math/pull/1343#discussion_r331448932